src/rust/engine/graph/src/lib.rs

```diff
@@ -584,17 +649,25 @@ impl<N: Node> Graph<N> {
   where
     C: NodeContext<Node = N>,
   {
-    // Initialize the state while under the lock...
-    let state = {
-      let mut inner = self.inner.lock().unwrap();
-      let id = inner.ensure_entry(EntryKey::Valid(node));
-      inner
-        .entry_for_id_mut(id)
-        .map(|entry| entry.state(context, id))
-        .unwrap_or_else(|| future::err(N::Error::invalidated()).to_boxed().shared())
-    };
-    // ...but only `get` it outside the lock.
-    state.then(unwrap_entry_res::<N>).to_boxed()
+    let mut inner = self.inner.lock().unwrap();
+    let id = inner.ensure_entry(EntryKey::Valid(node));
+    if let Some(entry) = inner.entry_for_id_mut(id) {
+      entry.get(context, id)
+    } else {
+      future::err(N::Error::invalidated()).to_boxed()
+    }
+  }
+
+  ///
+  /// When the Executor finishes executing a Node it calls back to store the result value. We use
+  /// the run_token value to determine whether the Node changed while we were busy executing it,
+  /// so that we can discard the work.
+  ///
```

Comment:
Really not trying to nitpick here: when would this happen? Would this occur basically only if this is a long-running Node with frequently-changing input? If I'm some object outside of the graph waiting on the result of some Node downstream of this one, would this mean that the value I *would* have received for the Node could be different depending on whether some other object happens to make a request for something else that transitively uses this node in the meantime?

Reply:
> Would this occur basically only if this is a long-running Node with frequently-changing input?

It would occur when while a Node is running, one of its transitive dependencies is invalidated, which will clear or dirty this Node, making the work that is executing irrelevant.

> would this mean that the value I would have received for the Node could be different depending on whether some other object happens to make a request for something else that transitively uses this node in the meantime?

No: requests to read things are "read only"/"append only" in terms of the graph: they only cause the graph to grow. `Graph::invalidate_from_roots` is what causes inputs to "change".

Reply:
Oh ok, awesome! I'm not clear on when one of a Node's transitive deps would be invalidated during the course of its execution unless there's more than one path from roots to the Node? Because I'm assuming a Node is only going to execute in the first place if one of its transitive dependencies is invalidated?

Reply:
And the file would be invalidated again if e.g. it is a source file and the user edits it after making a request for some product, but before the request completes, because watchman would note the file had changed and upon receiving that notif we would mark the corresponding snapshot node for the changed file as invalid?

Reply:
Corect,

Reply:
Ok great, I had a misconception about the interface from the engine to the graph, I think this method is probably fine as is.

Reply:
Also, note that this method is not public... it's only called from this module.